plugin/pulpcore/plugin/changeset/model.py

```diff
@@ -198,10 +298,51 @@ def save(self):
         Update the DB model to store the downloaded file.
         Then, save in the DB.
         """
-        if self.path:
-            self.model.file = File(open(self.path, mode='rb'))
-            self.model.downloaded = True
-        super(RemoteArtifact, self).save()
+        artifact = None
+
+        if not self.fetched:
+            if self._path:  # downloaded
+                try:
+                    with transaction.atomic():
+                        self.model = Artifact(
+                            file=File(open(self._path, mode='rb')),
+                            **self.monitor.dict())
+                        self.model.save()
+                except IntegrityError:
+                    q = self.artifact_q()
+                    self.model = Artifact.objects.get(q)
+                finally:
+                    artifact = self.model
+        else:
+            artifact = self.model
+
+        try:
+            with transaction.atomic():
+                content_artifact = ContentArtifact(
+                    relative_path=self.relative_path,
+                    content=self.content.model,
+                    artifact=artifact)
+                content_artifact.save()
+        except IntegrityError:
+            content_artifact = ContentArtifact.objects.get(
+                relative_path=self.relative_path,
+                content=self.content.model)
+            if self.fetched:
+                content_artifact.artifact = artifact
+                content_artifact.save()
+
+        try:
+            with transaction.atomic():
+                deferred_artifact = DeferredArtifact(
+                    url=self.url,
+                    importer=self.importer,
+                    content_artifact=content_artifact,
+                    size=self.model.size,
+                    **{f: getattr(self.model, f) for f in Artifact.DIGEST_FIELDS})
+                deferred_artifact.save()
+        except IntegrityError:
+            # TODO: update?
```

Comment:
Seems reasonable to me.

There is a potential use case for an importer deliberately doing this. Consider that the layout of a yum repo changed, but the same content is present. A smart importer would recognize this and update the URLs on any DeferredArtifact entries. Putting each such content unit in the iterator of additions would be a reasonable way to ensure it got updated in the DB.